File: src/graph/state.py
"""Agent state for the UP Police Data Analyst Agent's LangGraph query pipeline.

See spec/agent.md -> "Agent State" (binding). Every node reads/writes a subset
of these fields; the full shape travels through every node so LangGraph can
merge partial updates.
"""
from typing import TypedDict


class AgentState(TypedDict, total=False):
    # Identity
    query_run_id: str
    user_id: str
    dataset_id: str

    # Input
    question: str
    conversation_history: list[dict]  # [{"question": str, "answer": str}, ...] last N turns

    # Context (populated by load_context)
    schema_context: dict  # {columns: [...], row_count, date_range}
    sample_rows: list[dict]  # <=5 rows, free-text/narrative columns redacted
    annotations: dict[str, str]  # column_name -> business-rule note (empty {} until Phase 2)

    # Classification
    complexity: str  # "simple" | "complex"
    needs_clarification: bool
    clarifying_question: str | None

    # Planning (complex questions only)
    plan: list[str] | None

    # Code generation / execution loop
    current_code: str | None
    attempts: list[dict]  # [{attempt_number, code, stdout, result, error, duration_ms}]
    retry_count: int
    max_retries: int
    execution_result: dict | None
    execution_error: str | None

    # Output
    final_answer: str | None
    key_numbers: dict | None
    assumptions: list[str]
    followups: list[str]  # [] in Phase 1 (stub); real from Phase 3

    # Usage
    prompt_tokens: int
    completion_tokens: int
    estimated_cost_usd: float

    # Control
    current_node: str  # written after every node for progress polling
    status: str  # "pending" | "needs_clarification" | "completed" | "failed"
    error: str | None  # set by any node on fatal failure

File: src/agent/state.js
"use strict";

/**
 * LangGraph.js state channel definitions for the agent graph.
 * Mirrors spec/agent.md's `AgentState` shape exactly.
 *
 * IMPORTANT (observed langgraph@0.2.x behaviour): `default` initializers are
 * NOT eagerly merged into the object handed to the first node — a node
 * invoked with a sparse initial input can still see `undefined` for fields
 * with declared defaults. Every node in src/agent/nodes/*.js therefore reads
 * state defensively (e.g. `state.attempts || []`) rather than relying solely
 * on the declared default.
 */

const { Annotation } = require("@langchain/langgraph");

const AgentState = Annotation.Root({
  // Identity
  runId: Annotation(),
  userId: Annotation(),
  datasetId: Annotation(),

  // Input
  question: Annotation(),
  conversationHistory: Annotation({ default: () => [] }), // Phase 2

  // Dataset context (populated by load_context)
  datasetProfile: Annotation(),
  annotations: Annotation({ default: () => [] }), // Phase 2
  sampleRows: Annotation({ default: () => [] }), // <=20 rows, LLM prompt context ONLY
  rows: Annotation({ default: () => [] }), // full parsed dataset — in-process only
  columns: Annotation({ default: () => [] }),

  // Classification / planning
  // NOTE: named `planSteps` (not `plan`, per the spec/agent.md field name)
  // because @langchain/langgraph 0.2.x forbids a state channel and a graph
  // node from sharing the same identifier, and the `plan` NODE name is the
  // one that must match spec/agent.md's graph topology exactly.
  complexity: Annotation(),
  clarifyingQuestion: Annotation({ default: () => null }),
  planSteps: Annotation({ default: () => [] }),

  // Code generation / execution loop
  generatedCode: Annotation(),
  codeExplanation: Annotation(),
  executionResult: Annotation({ default: () => null }),
  attempts: Annotation({ default: () => [] }),
  retryCount: Annotation({ default: () => 0 }),

  // Inspection
  inspectionVerdict: Annotation(),
  inspectionNote: Annotation(),

  // Output
  answer: Annotation({ default: () => null }),
  keyNumbers: Annotation({ default: () => [] }),
  assumptions: Annotation({ default: () => [] }),
  chartSpec: Annotation({ default: () => null }), // Phase 2
  followups: Annotation({ default: () => [] }), // Phase 1: always []
  tokenUsage: Annotation({ default: () => ({ promptTokens: 0, completionTokens: 0 }) }),

  // Control
  error: Annotation({ default: () => null }),
  status: Annotation({ default: () => "running" }),
});

module.exports = { AgentState };
